common/config/fx.go

```diff
@@ -0,0 +1,58 @@
+// The MIT License
+//
+// Copyright (c) 2020 Temporal Technologies Inc.  All rights reserved.
+//
+// Copyright (c) 2020 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package config
+
+import (
+	"go.uber.org/fx"
+
+	"go.temporal.io/server/common/primitives"
+)
+
+var Module = fx.Provide(
+	provideRPCConfig,
+	provideMembershipConfig,
+	provideServicePortMap,
+)
+
+func provideRPCConfig(cfg *Config, svcName primitives.ServiceName) *RPC {
+	c := cfg.Services[string(svcName)].RPC
+
+	return &c
+}
+
+func provideMembershipConfig(cfg *Config) *Membership {
+	return &cfg.Global.Membership
+}
+
+func provideServicePortMap(cfg *Config) ServicePortMap {
+	servicePortMap := make(ServicePortMap)
+	for sn, sc := range cfg.Services {
+		servicePortMap[primitives.ServiceName(sn)] = sc.RPC.GRPCPort
+	}
+
+	return servicePortMap
+}
+
+type ServicePortMap map[primitives.ServiceName]int
```

Comment:
I have mixed feelings about this kind of super-fine-grained fx usage, especially for stuff that has only one consumer. It's sort of nice that things just magically get called and compose, but it's also.. well, magic, and hard to trace through (using standard IDE tools that I know of). I'd like to get some other opinions before we go all-in on this sort of pattern

Reply:
Or @alexshtin

Reply:
My argument goes like this:
1. To permit other `map[primitives.ServiceName]int` objects in the graph, we would need to add tags like `fx.Annotate(provideServicePortMap, fx.ResultTags("name:\"servicePortMap\"))` which seems verbose and error-prone.
2. These fine-grained types actually increase traceability because it's very easy to jump to references to `ServicePortMap`, as opposed to searching for `map[primitives.ServiceName]int`.
3.  I think that, since the only dependency for this object is a config, this provider should live here so that anyone can easily get a service port map from the config
4. If other packages are creating the service port map from the config, they would violate the Law of Demeter because they would have to access properties deep within the structure of the config.

Reply:
But this thing has a single user, no one else needs a service port map. You can just build it in `getMonitor` right before you call `newMonitor`, there's no reason for it to be visible to fx at all. It's just plain old data, there's no need to enforce uniqueness and share it

Reply:
This will be reused soon by new membership implementations.

Reply:
Ah! You should have said that earlier :)  In the future if you have stacked PRs ready to go, just send them all (as drafts or whatever), it's a lot easier for the reviewer if they can see where things are going / how things will be used